tests/nn/data_parallel/test_fsdp_optimizer_utils.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+import functools
+from time import time
+
+from parameterized import parameterized
+import torch
+from torch.optim import SGD, Adadelta, Adam  # type: ignore
+
+from fairscale.nn import FullyShardedDataParallel
+from fairscale.optim.utils import recursive_copy_to_device
+from fairscale.utils.testing import objects_are_equal
+
+from .test_fsdp import (
+    DistributedTest,
+    DummyProcessGroup,
+    NestedWrappedModule,
+    TransformerWithSharedParams,
+    rename_test,
+    spawn_and_init,
+)
+
+
+def first_tensor_shape(dct):
+    for k, v in dct.items():
+        if torch.is_tensor(v):
+            return v.numel()
+    return 0
+
+
+def assert_equal(a, b):
+    assert a == b, f"{a} != {b}"
+
+
+class TestOptimizerUtils(DistributedTest):
+    @parameterized.expand(
+        [[functools.partial(SGD, momentum=0.9), True], [SGD, False], [Adam, False], [Adadelta, True]],
+        name_func=rename_test,
+    )
+    def test_consolidate_optimizer(self, optim_fn, transformer):
+        config = {"mixed_precision": True, "flatten_parameters": True}
+        test_fn = functools.partial(
+            self._test_consolidated_optimizer, config, optim_fn=optim_fn, transformer=transformer
+        )
+
+        spawn_and_init(test_fn, world_sizes=[min(torch.cuda.device_count(), 4)])
+
+    @classmethod
+    def _test_consolidated_optimizer(self, config, rank, group, optim_fn=torch.optim.SGD, transformer=False):
+        """FSDP.gather_full_optim_state_dict() should return something very similar to optimizer.state_dict()"""
+        # Establish reference behavior.
+
+        if transformer:
+            fsdp = self.get_wrapped_model(group, config=config).cuda()
+            unwrapped_model = TransformerWithSharedParams(group).cuda()
+        else:
+            fsdp = FullyShardedDataParallel(NestedWrappedModule(group, wrapper_config=config), group, **config).cuda()
+            unwrapped_model = NestedWrappedModule(group, wrapper_config=None).cuda()
+
+        try:
+            fsdp_optim = optim_fn(fsdp.parameters(), lr=0.01,)
+            optim_unwrapped = optim_fn(unwrapped_model.parameters(), lr=0.01)
+        except TypeError:  # AdaScale
```

Comment:
do you actually mean "AdaScale" here? I don't see AdaScale being used here in this test.

Reply:
yes, nice catch